queues.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+// A queue must be read and write
+type Queue interface {
+	send(DomainEntry) error
+	pull(int) ([]DomainEntry, error)
+	peek(string) int
+	persist() error
+}
+
+type cachedQueue struct {
+	store   map[string]int
+	backend Backend
+}
+
+func NewCachedQueue(persistanceBackend Backend) *cachedQueue {
+	q := new(cachedQueue)
+	q.store = make(map[string]int)
+	q.backend = persistanceBackend
+	return q
+}
+
+func (q cachedQueue) send(entry DomainEntry) error {
+	stats := q.store[entry.Domain]
+	if stats == -1 {
+		return nil
+	} else if !entry.IsBounce && stats >= 0 {
+		if stats >= 1000 { // a magic number
+			q.store[entry.Domain] = 0
```

Comment:
Would you have enough memory to store millions of different domains? What if an OOM killer comes along?

Reply:
It truthfully depends on the host that is running the process.

A semi guess would be that it would be quite impressive for the local cache to be that large. The DomainEntry is at max a   KB (253 * 2 bytes) + 2 DWORDs at the most suboptimal data representations + padding and stuff. If it is a problem I think the worker could possibly trunc stale records.

Reply:
KB * 10,000,000 = 10Gb... or I'm missing something?

Reply:
That is indeed a valid concern in the worst case and possibly even the average case. I do not know the stats of the data set beyond anecdotal knowledge. 

The nodes should attempt to offload their storage to cassandra whenever it is appropriate. Something like a push of the data to the backend at a critical point with a local flush of cleared values. A bit of a bounded FIFO queue?

The map keys could be compressed to reduce storage. I'm not sure of how effective a hash of the domains would be at ensuring uniqueness (sha, etc). That would be an excellent option if it proves to maintain that.